GeoffreyBooth-proposal.md

```diff
@@ -0,0 +1,78 @@
+
+### Developer Happiness: Supporting Optional ES2015+ Features
+
+Features that developers enjoy in ES2015 that are appropriate to implement in CoffeeScript, like `const`, should be implemented as time permits. These features should be implemented in a new compiler, assuming we build one; and possibly also in the old compiler too if such an implementation is easy and doesn’t break backward compatibility. Any feature implemented in both compilers should have identical syntax.
+
+### Modernizing Output
+
+Lastly, CoffeeScript’s output should be modernized as much as possible. Fat arrows should output as fat arrows, destructuring should output as destructuring, etc. But I can’t emphasize enough that this should be the **last priority**—it makes our source code cleaner, and debugging eventually easier when ES2015 features are supported natively in runtimes, but otherwise ES2015 output has little benefit for developers. That said, *new* features, especially ES2015+ ones, should definitely be output as ES2015+ code.
+
+## A New Compiler
+
+The current CoffeeScript 1.x compiler is extremely well commented, but it really needs those comments. Understanding it will bend your brain into knots. The original goal of [CoffeeScript Redux](https://github.com/michaelficarra/CoffeeScriptRedux) was to refactor CoffeeScript’s codebase, not to add new features.
+
+Don’t underestimate how hard it is to write your own language compiler. If it were easy, CoffeeScript Redux would have been finished; as it stands, that project was essentially abandoned a year ago.
+
+Also don’t underestimate how hard it is to add new features to the CoffeeScript 1.x compiler. If that were easy, someone else would’ve added modules and classes to it by now.
+
+The CoffeeScript community suffers from too many people with a lot of passion but little time to invest in development. For that reason, I consider a new-from-scratch compiler to be off the table. We’ll just end up with another Redux.
+
+So we have only two options for starting points for our new compiler:
+
+* The current CoffeeScript 1.x compiler.
+	* Pros: already supports all existing CoffeeScript code.
+	* Cons: complicated, difficult-to-understand codebase.
+* The CoffeeScript Redux compiler, probably from [this PR](https://github.com/michaelficarra/CoffeeScriptRedux/pull/344).
+	* Pros: better codebase, and the PR already supports many ES2015 features.
+	* Cons: we inherit Redux’s issues and unfinished work, including any incompatibilities with existing CoffeeScript code.
+
+I don’t have a recommendation as to one or the other. I invite contributors to both projects to chime in.
```

Comment:
I don’t feel qualified to make a choice between them, since I’ve only just recently started contributing to the main repo and I’ve never worked in the Redux repo. Someone like @michaelficarra or @lydell needs to chime in with some more detail, and either tell us which way to go or give us enough detail that we can decide as a community.